### DISCUSSION TITLE
Lighthouse Geometry Estimation error

### DISCUSSION THREAD

**Discussion by LucasTheIntern** [bitcraze/discussions#1184] [Q&A] [answered]
I have two lighthouse V2 base stations, a crazyflie 2.1, and lighthouse v4 deck. I'm trying to set everything up to fly within a cube that's approximately 5'x5'x5'. I have had this working in the past, but it has stopped working after being disassembled/moved/reconstructed. Photo of setup:
![image](https://github.com/bitcraze/discussions/assets/120214901/5b3a0645-54de-4b73-b0de-bf5b35ce5ee6)
The base stations are mounted in opposite corners at the top of the enclosure. Previously, this provided sufficient coverage of the flying space. The enclosure was only moved a few feet, so the environment hasn't really changed. 

After the enclosure was reassembled, the crazyflie was unable to maintain stable flight using the buttons on the CF Client Flight Control tab (running on VM). It would take off, and then start swaying about in an increasingly uncontrolled manner until hitting something or flipping upside down.

I figured that this had something to do with the base stations not being _excactly_ where they were before, so I tried to redo the geometry estimation. For whatever reason, the copy of the VM that I have been working with (newest version, some modifications) wouldn't let me open the GUI for geometry estimation. I created a new copy of the VM and was able to open the GUI. I am now able to get through all of the steps provided in the GUI, but the geometry estimation always fails:
![image](https://github.com/bitcraze/discussions/assets/120214901/80085268-6d9f-4547-b712-0ff2d151f716)
Note that the base stations appear to be very close to one another in this screenshot. This was not the case when I first started my troubleshooting, but I'm not sure when it started.

Here are some ideas I have tried:

- Swapping Lighthouse decks
- moving base stations
- resetting base station channels
- using an older version of the VM
- using a clean copy of the newest version of the VM
- Restarting everything (power cycling base stations, crazyflie, VM)
- Turning off all lights in the lab, covering all reflective surfaces (didn't have to do this before)
- Reflashing firmware to the crazyflie and lighthouse deck

**Other info:**

The "Basestation Status" area on the Lighthouse Positioning tab looks like this after connecting:
![image](https://github.com/bitcraze/discussions/assets/120214901/9b022419-44c6-4a37-944e-c9846ae6ed1c)
I'm not sure what the blue region means, I don't remember seeing that before.

The plotter tab in the clean VM doesn't show anything when I select "lhStatus", though the older version of the VM does show a few flat lines (But not all lines shown in the legend). I wanted to check this after reading this post:
https://github.com/orgs/bitcraze/discussions/1083
But swapping out the LH deck with a fresh one didn't work for me, and I doubt that I have two faulty decks.

The LED indicators on the lighthouse deck and base stations seem to be working just fine, and change when the sensors are covered by my hand. (both solid green with line of sight, deck turns orange when blocked, base stations flash blue when blocked).

I do observe a jittery horizon in the Flight Control tab after booting the crazyflie as described in this unanswered post with a similar issue:
https://github.com/orgs/bitcraze/discussions/118
This jittering usually calms down after several minutes of waiting (which I have to do anyway before the client will respond to user input)

I did have a temporary issue with the VM claiming that no sweeping angles were detected, even though the indicator on the lighthouse deck was green. Resetting the base station channels fixed this.

I've attached the output from the console tab in a file "console.txt".
[console.txt](https://github.com/bitcraze/discussions/files/14284497/console.txt)

I've attached the output from the terminal window that opens with the client in a file "terminal.txt".
[terminal.txt](https://github.com/bitcraze/discussions/files/14284495/terminal.txt)

Any ideas?

Thanks, 
Lucas

**Comment by ataffanel** [comment]
Hi,

Thanks for the detailed question and sorry for the late answer. I tried to reproduce your problems and to me the main problem is the VM. Connecting a crazyflie from within the VM and configuring it seems much worse than before, something has clearly deteriorated with time (virtualbox is now using Microsoft's virtualization, maybe that is a reason ...).

Anyway, I think one way forward for you would be to install the client on the machine directly. I worked at fixing the client for python 3.12 on windows and this is now released as version 2014.2.1 of the client. You can now install python from the windows store and then type ```pip install cfclient``` in a console. The client can then be launched with ```cfclient``` or ```python3 -m cfclient.gui```. I checked that this works on windows to setup a ligthouse system. Of course if you already have python on this machine you can use your existing python.

We will have a look at the state of the VM to find out what did happen and how we can go forward with it since the current state is unfortunately not usable anymore.

**Reply by LucasTheIntern** [reply]
Thanks for the reply, Arnaud!

I tried this just now, with the same result as before ("No lighthouse base station geometry solution could be found!"). 

There was no text output (or instructions) in the "manage geometry" GUI. Otherwise, running the client on my windows 11 machine without the VM was a superior experience to working within the VM! Much faster. 

The lighthouse positions shown in the client are still nearly on top of each other, which I suppose is because the old geometry is still stored on the crazyflie.

Here is the output in the terminal (Anaconda Prompt):
```
(host_crazyflie) C:\Users\fiddl>cfclient
UserWarning: Using SDL2 binaries from pysdl2-dll 2.24.0
INFO:cfclient.gui:Disabling STL printouts
INFO:cfclient.utils.input.inputreaders:Input readers: ['linuxjsdev', 'pysdl2']
INFO:cfclient.utils.input.inputreaders:Could not initialize [linuxjsdev]: Only supported on Linux
INFO:cfclient.utils.input.inputreaders:Successfully initialized [pysdl2]
INFO:cfclient.utils.input.inputinterfaces:Found interfaces: ['leapmotion', 'wiimote', 'zmqpull']
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [leapmotion]: Leap Motion library probably not installed (No module named 'leapsdk')
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [wiimote]: Missing cwiid (wiimote) driver No module named 'cwiid'
INFO:cfclient.utils.config:Dist config read from C:\Users\fiddl\anaconda3\envs\host_crazyflie\lib\site-packages\cfclient/configs/config.json
INFO:cfclient.utils.config:Config file read from [C:\Users\fiddl\AppData\Local\Bitcraze\cfclient/config.json]
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [zmqpull]: ZMQ input disabled in config file
INFO:cfclient.utils.zmq_led_driver:ZMQ led disabled in config file
INFO:cfclient.utils.zmq_param:ZMQ param disabled in config file
INFO:cfclient.utils.input:Using device blacklist [(VirtualBox|VMware)]
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cflib.crtp.radiodriver:v3.2 dongle with serial N/A found
INFO:cfclient.utils.input.inputreaders.pysdl2:Looking for devices
INFO:cfclient.utils.input.inputreaders.pysdl2:Found 1 devices
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.utils.input:Selected MUX: Normal
INFO:cfclient.ui.main:Role of Xbox 360 Controller is Device
INFO:cfclient.utils.input.mux:Adding device Xbox 360 Controller to MUX Normal
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
INFO:cflib.crtp.radiodriver:v3.2 dongle with serial N/A found
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
INFO:cflib.crazyflie:Callback->Connection initialized[radio://0/80/2M]
INFO:cflib.crazyflie:We are connected[radio://0/80/2M], request connection setup
INFO:cflib.crazyflie.platformservice:Request _request_protocol_version()
INFO:cflib.crazyflie:Callback->Connected to [radio://0/80/2M]
WARNING:cflib.crazyflie.log:Error no LogEntry to handle id=8
INFO:cflib.crazyflie.platformservice:_crt_service_callback
INFO:cflib.crazyflie.platformservice:Request protocol version
INFO:cflib.crazyflie.platformservice:_platform_callback
INFO:cflib.crazyflie.platformservice:Protocol version (platform): 7
INFO:cflib.crazyflie.toc:TOC for port [5] found in cache
INFO:cflib.crazyflie:Log TOC finished updating
INFO:cflib.crazyflie.mem:10 memories found
INFO:cflib.crazyflie:Memories finished updating
INFO:cflib.crazyflie.toc:TOC for port [2] found in cache
INFO:cflib.crazyflie:Param TOC finished updating
INFO:cflib.crazyflie:Callback->Connection setup finished [radio://0/80/2M]
INFO:cfclient.utils.logconfigreader:Parsing [stabilizer.json]
INFO:cfclient.ui.main:LED write done callback
INFO:cflib.crazyflie.log:Have successfully started logging for id=1
INFO:cflib.crazyflie.log:Have successfully started logging for id=3
INFO:cfclient.ui.tabs.LEDTab:Memory: id=4, type=LED driver, size=24
INFO:cflib.crazyflie.log:Have successfully started logging for id=4
INFO:cflib.crazyflie:All parameters updated
INFO:cflib.crazyflie:Callback->Connection completed [radio://0/80/2M]
INFO:cflib.crazyflie.log:Have successfully started logging for id=5
False
False
True
True
True
C:\Users\fiddl\anaconda3\envs\host_crazyflie\lib\site-packages\cflib\localization\lighthouse_types.py:61: ComplexWarning: Casting complex values to real discards the imaginary part
  return Pose(Rotation.from_quat(R_quat).as_matrix(), t_vec)
No lighthouse base station geometry solution could be found!
```

**Reply by ataffanel** [reply]
Hi,

Can you precise what you mean by "There was no text output (or instructions) in the "manage geometry" GUI."? Does the GUI looks like the documentation or does there seems to be a graphical bug in the client? The log you sent seems to imply some graphical bugs.

The Crazyflie needs to move in the space during calibration and be as much of possible in view of both basestation, is that the case during your attempts?

**Reply by LucasTheIntern** [reply]
Hi Arnaud, sorry about the extremely late reply... got sidetracked with a bunch of other issues.

The geometry estimation GUI was showing the diagrams and buttons, but wasn't giving any kind of feedback like what I saw in the same GUI using the VM. Geometry estimation with the VM provided text in the GUI regarding quality of the measurements, whether they needed to be redone, etc. 

I just tried again using the PC client on a new windows machine. I also just now updated the client. When I open the "Manage Geometry" window and select "Estimate Geometry," this is what I get:
![image](https://github.com/user-attachments/assets/dc895226-4656-4782-8068-55cdd54e6699)

That's what I was talking about before. There's no text in the GUI. 

Clicking "Start Measurement" causes the expected delay, and then the button changes to "Restart Measurement" and the "next" button becomes clickable. The same is true for all of the next few steps. 

If this still isn't clear, I can try recording a video!

**Reply by knmcguire** [reply, marked as answer]
Ah yes that is a known bug that is currently not fixed yet. There is a proposed solution here: https://github.com/bitcraze/crazyflie-clients-python/issues/691

**Reply by LucasTheIntern** [reply]
Thanks Kimberly! 

I was able to complete the process successfully after switching to the dark blue mode! 

I guess something between the new host machine and updated PC client fixed the other issues. The LH system works as expected now.

**Reply by knmcguire** [reply]
awesome! glad to have helped.

I think we should remove that theme switching part of the code for the next release, but at least you are unblocked :D
